### ISSUE TITLE
Docker: replace Chromium for drawing graphs ?

### ISSUE THREAD

**Issue by bschmalhofer** [RotherOSS/otobo#3550]
The lightweight alternative to Chromium is to use node.js on the server. For examples how this can be done see:

- https://medium.com/@jkrogmann/wie-du-mit-d3-js-und-node-js-charts-erstellst-und-als-png-abspeicherst-5f2b3e132bc2
- https://www.jafaraziz.com/blog/transform-geojson-to-png-with-d3-js/
- https://www.npmjs.com/package/d3-node (looks like this is not needed)
- https://medium.com/@palaniscit/bar-chart-using-node-js-and-d3-d0cb08fbe8f6 (using d3-node)

_Originally posted by @bschmalhofer in https://github.com/RotherOSS/otobo/issues/3542#issuecomment-2194235568_

**Comment by bschmalhofer** [comment]
This needs to be discussed. The scope might be larger than just drawing some graphs, other items might have to be made printable as well.

**Comment by bschmalhofer** [comment]
The implementation for drawing the charts is in  _var/httpd/htdocs/js/Core.UI.AdvancedChart.js_ . That library uses the NVD3 library, see https://github.com/nvd3/nvd3 or one of the forks. Development seems to have stalled for this JS library.

Server side generation of graphs might be an option too.

**Comment by bschmalhofer** [comment]
The decision was to concentrate on server side generation of charts. Neither a browser nor node.js on the server are to be used. PDL should not be required.

The first task is to identify a viable solution for generating charts on the server. Herbert Breuning, lichtkind, kindly collected information about [Plotten mit Perl](https://lichtkind.de/vortrag/folien/plot5.pdf).

- [Chart](https://metacpan.org/dist/Chart): Solid 90s look, uses GD for PNG generation
- [Chart::Plot](https://metacpan.org/pod/Chart::Plot): abandoned
- [GD::Graph](https://metacpan.org/pod/GD::Graph) often first search hit, in maintenance mode
- PDL::Graphics::*   PDL should not be added to OTOBO
- [Graphics::PLPlot](https://metacpan.org/pod/Graphics::PLplot) abandoned
- [Chart::Plotly](https://metacpan.org/pod/Chart::Plotly) very nice, needs chartly.js, needs JavaScript
- [Chart::Kaleido::Plotly](https://metacpan.org/pod/Chart::Kaleido::Plotly) for downloading static charts
- [Chart::GGPlot](https://metacpan.org/pod/Chart::GGPlot) Very nice charts, needs PDL, maybe only Chartly backend, which is JS
- [SVG::Graph](https://metacpan.org/pod/SVG::Graph) not maintained
- [SVG::TT::Graph::Bar](https://metacpan.org/pod/SVG::TT::Graph::Bar) last release 2020,
- [Chart::Gnuplot](https://metacpan.org/pod/Chart::Gnuplot) a wrapper for **Gnuplot, a safe bet**, the license should be fine

My favourite:

[Gnuplot::Builder](https://metacpan.org/pod/Gnuplot::Builder) is future save because it simply wraps Gnuplot. It seems to have a nice interface and seems to be maintained. Spawns a Gnuplot process, which is fine. Looks good enough: http://www.gnuplot.info/demo_4.6/histograms.html

**Comment by stefanhaerter** [comment]
Seeing this, I was reminded about Herbert Breunungs talk [Plotting mit Perl](https://act.yapc.eu/gpw2022/talk/7780) (Plotting with Perl) at the German Perl and Rakulang Workshop 2022. Sadly, I couldn't find the recording on YouTube.

**Comment by bschmalhofer** [comment]
Yes, that is the talk I mined for information. The slides are linked above.

**Comment by stefanhaerter** [comment]
Ah, sorry - I didn't read properly 🙈

**Comment by bschmalhofer** [comment]
The first results are not really satisfying. Propably need to fiddle with the settings some more. Maybe the resizing of the .png has problems as well. 

<img width="1908" height="1055" alt="Image" src="https://github.com/user-attachments/assets/b787e3f1-27d6-4b43-89d9-776620b08a33" />

**Comment by bschmalhofer** [comment]
A bit nicer after some fiddling:

<img width="2670" height="1477" alt="Image" src="https://github.com/user-attachments/assets/3def13b6-2327-4379-92b4-45833a9594bd" />

TODO:

- [x] Try generating vector graphics, maybe the terminal pdfcairo.

**Comment by bschmalhofer** [comment]
Looks better with the terminal pdfcairo, especially when zooming in.

<img width="1853" height="896" alt="Image" src="https://github.com/user-attachments/assets/c32aa89d-a00c-4024-bb69-068b17251016" />

TODO:

- [x] Stacked output, just to have something that the report with Chromium does not provide
- [ ] Adapt the size to use the full width of the page
- [ ] Place the chart between Preface and Postface
- [ ] Size the xtics, add newline for the odd ticks, to make it look like layout on two lines
- [x] Port D3::LineChart
- [ ] Port D3::StackedAreaChart, tentatively implemented, needs to be verified

**Comment by bschmalhofer** [comment]
Here is the line chart:

<img width="1145" height="648" alt="Image" src="https://github.com/user-attachments/assets/8aa7aa37-8b68-4614-bc78-d6c40724561f" />

**Comment by stefanhaerter** [comment]
Looks pretty neat already.

**Comment by bschmalhofer** [comment]
First stab at a stacked area chart:

<img width="1853" height="896" alt="Image" src="https://github.com/user-attachments/assets/5c7fc0fb-28d9-49ec-ae1b-632382bddd6f" />

**Comment by bschmalhofer** [comment]
Here is the latest complete report. It contains grapsh created with D3 and Gnuplot for comparison:

[Welcome to OTOBO.pdf](https://github.com/user-attachments/files/32463568/Welcome.to.OTOBO.pdf)

**Comment by bschmalhofer** [comment]
Here is the latest complete report. It contains graphs created with D3 and Gnuplot for comparison:

[Welcome to OTOBO.pdf](https://github.com/user-attachments/files/32465399/Welcome.to.OTOBO.pdf)

**Comment by bschmalhofer** [comment]
Worked on the layout, positioning and sizing.

Ran into an issue where plots were not visible when debut output with `Dx` was deactivated. Maybe a timing issue.

**Comment by bschmalhofer** [comment]
When working on support for Gnuplot I noticed that the OTOBO PDF generation uses some deprecated methods. This is fine, as these methods are still supported for compatability. However, it is nicer to use the current methods.

One such adaption is already committed in rel-11_1: https://github.com/RotherOSS/otobo/commit/b38db089da738b6728dd66517594133b2f4cfb35 .

**Comment by bschmalhofer** [comment]
Here is the current version of the sample report:

[Welcome to OTOBO.pdf](https://github.com/user-attachments/files/32569095/Welcome.to.OTOBO.pdf)

**Comment by bschmalhofer** [comment]
No Unicode problems surfaced at the first glance, Defined the ticket state "ü ue, ⛄ snowman"  and it was correctly shown in both, D3 and Gnuplot, types of charts.
